packages/dds/tree/docs/wip/schema-evolution.md

```diff
@@ -0,0 +1,583 @@
+   * To mitigate that problem, another option would be to have the "Unknown" type be automatic - or "opt-out" rather than "opt-in".
+     That is, all applications must handle the "Unknown" type in every field, unless they intentionally make the decision not to.
+   * A third approach would be to automatically filter out unrecognized node types where applicable.
+     In the example above, the original client would simply not be able to observe that there are Image nodes in the canvas field - the field would appear shorter than it does to the upgraded clients if there are images in it.
+     This might be appropriate for sequence fields, but would not work for required fields because a required field must not appear empty.
+
+     > TODO: This is one of Noah's off-the-cuff ideas. Whether or not filtering makes sense for optional fields (or at all) is debatable.
+
+     This strategy could be used in combination with the "Unknown" type - if Unknown is in the field's set of allowed types, then the foreign type will present as an Unknown, otherwise it will be filtered.
+
+1. **Remove an allowed type.**
+   This scenario can employ the same strategy as "Add an allowed type" above, namely, the "Unknown" catch-all type.
+   The newer schema can simply add the "Unknown" type to the set of allowed types if it doesn't already exist.
+   Then the removed type, when present in the document, will present as an "Unknown".
+
+   This does not require the developer to plan ahead - "Unknown" can be added to the new view schema when the field is first removed, rather than having to be added preemptively in the original schema as is the case when adding an allowed type.
+   However, the _stored_ schema may not be upgraded to remove the node type until the number of clients with the new view schema has passed the saturation point.
+   This is because so long as an old client exists, it might try to write the removed type into the document, and if the stored schema has already removed that type then they will experience an error.
+
+1. **Change the type of a node.**
+   This can be accomplished by first removing the node type and then adding a new one, according to the processes outlined above.
+   Therefore it takes on the drawbacks of both - the "Unknown" type must be added preemptively, and the stored schema may not be upgraded until the new clients have reached saturation.
+
+   In the event that the developer merely wants to rename the _identifier_ of a node, but otherwise keep the node the same, we can support that with an aliasing feature.
+   Specifically, when defining the node in the view schema it can be supplied with both a "view identifier" (the new name) and a "stored identifier" (the existing name).
+   The view identifier will show up in the source code when reading/writing the tree, but it will be automatically translated to the legacy stored identifier when reading/writing to the document.
```

Comment:
Can this be generalized? I think you describe any scenario that can be solved by bi-directional adapter (we continue to write in old schema, I assume forever).